File: 110. Balanced Binary Tree/fault.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @description https://leetcode.com/problems/balanced-binary-tree/
 * @type self
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
  function dfs(node) {
    if (!node || (!node.left && !node.right)) {
      return true;
    }
    if (!node.left) {
      if (!(node.right.left === null && node.right.right === null)) {
        return false;
      }
    }
    if (!node.right) {
      if (!(node.left.left === null && node.left.right === null)) {
        return false;
      }
    }
    if (Math.abs(getHeight(root.left) - getHeight(root.right)) > 1) {
      return false;
    }
    return dfs(node.left) && dfs(node.right);;
  }
  return dfs(root);
};

function getHeight(root) {
  if (!root) {
    return 0;
  }
  return Math.max(getHeight(root.left), getHeight(root.right)) + 1;
}

File: 572. Subtree of Another Tree/fault.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
/**
 * https://leetcode.com/problems/subtree-of-another-tree/
 * self
 */
class Solution {
  public boolean isSubtree(TreeNode root, TreeNode subRoot) {
    Queue<TreeNode> queue = new LinkedList<TreeNode>();
    queue.offer(root);
    queue.offer(subRoot);
    
    while (!queue.isEmpty()) {
      TreeNode m = queue.poll();
      TreeNode n = queue.poll();
      if (m == null) return false;
      if (isSameTree(m, n)) return true;
      queue.offer(m.left);
      queue.offer(n);
      queue.offer(m.right);
      queue.offer(n);
    }

    return false;
  }

  public boolean isSameTree(TreeNode p, TreeNode q) {
    if (p == null && q == null) return true;
    if (p == null || q == null) return false;
    if (p.val != q.val) return false;
    return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
  }
}